README.md
twitter_autoload
================

Automatically loads new tweets to whatever feed you are watching

Ever find yourself keeping a browser tab of twitter open on your secondary screen, wishing that it would just get those new tweets and not require you to click to see them?
Well that (1st world) problem is now in the past. By installing this extension, your twitter feed will never again fall behind the time. As long as your window is scrolled to the top and your not in the middle of writing a tweet yourself, regardless if your on the front page, watching a search result, or a specific person, new tweets will automatically be loaded on to your feed.


https://chrome.google.com/webstore/detail/twitter-autoload/kjiihncojgddjdaeclofdpikkbandnke


CHANGELOG
Changelog:

1.0.1.2
- Fixed: Loading at the actual top of timeline, regardless of all the clutter twitter decides to put on top
- TODO: Figure out a more obvious way to pause updating then focusing a text area incase the user wants to do something above the timeline

1.0.1.1
- Fixed: Made internal timer to reset immediately instead of after next update when new tweets are found
- Prevent too many simultaneous requests

1.0.1.0
- Load new tweets when scrolled at or above top of timeline (tweets area)
- Less spammy updates: still keeps up with busy topics
- Now also watches for twitters own load new tweets bar, useful for feeds with less frequent content

1.0.0.2
- New punchline

1.0.0.1
- Removed useless permissions from manifest

1.0.0.0
- Stable & Feature complete.

0.0.0.6
- actually trigger event every time if visible

0.0.0.5
- only take contenteditable="true" in to account

0.0.0.4
- trigger updates trough uiShortcutGotoTopOfScreen instead of emulating a '.' press
- Spam updates less frequently
- Update while not visible, but less frequently
- Determine the intention to write by watching the contenteditable attribute instead of a specific class.

0.0.0.3
- Only loads tweets when tab is "visible". This update requires Chrome version 33.0+
https://developers.google.com/chrome/whitepapers/pagevisibility

0.0.0.2
- Initial release

twitter_autoload/autoload.js
/*! Twitter autoload source | (c) 2016 William Wendell */

var actualCode = "(function(){ 'use strict'; var Autoload = function(){ this.uq=false; this.v=!document.hidden; this.exp = -1; this.to = false; this.o = new MutationObserver((function(m){ this.undelay(); }).bind(this)); this.o2 = new MutationObserver((function(m){ this.undelay(); }).bind(this)); this.oc = { attributes: false, childList: true, characterData: false }; this.tl = document.getElementById('stream-items-id'); this.t0 = 2500; document.addEventListener('visibilitychange', (function(ev){ if (this.v=!document.hidden && this.uq) { this.tk(); } }).bind(this), false ); this.o.observe(document.querySelector('.js-new-items-bar-container'), this.oc); this.o2.observe(this.tl, this.oc); this.keepup(); }; Object.assign(Autoload.prototype, { su: function(px) { function getOffsetSum(e) { var top=0, left=0; while(e) { top = top + parseInt(e.offsetTop); left = left + parseInt(e.offsetLeft); e = e.offsetParent; } return {top: top, left: left}; } return !document.querySelectorAll('input:focus,textarea:focus,[contenteditable=\"true\"]:focus').length && window.pageYOffset - document.documentElement.clientTop - getOffsetSum(this.tl).top < px && this.v; }, tk: function() { if (this.su(10)) { document.dispatchEvent(new CustomEvent('uiShortcutGotoTopOfScreen')); this.uq = false; } else { if (this.v) { window.setTimeout( this.tk.bind(this), 100); } else { } } }, up: function() { if (!this.uq) { this.uq = true; this.tk(); return true; } else { return false; } }, undelay: function(f) { this.exp = -1; window.clearTimeout(this.to); this.to = window.setTimeout(this.keepup.bind(this), 100 ); }, delay: function() { if (this.exp < 10) { ++this.exp; } }, keepup: function() { if (this.up()) { this.delay(); } this.to = window.setTimeout(this.keepup.bind(this), parseInt(this.t0 * Math.pow(1.3, this.exp)) ); } }); window.onload = window.setTimeout(new Autoload(), 5000); })();";
var script = document.createElement('script');
script.appendChild(document.createTextNode(actualCode));
(document.head || document.documentElement).appendChild(script);
script.parentNode.removeChild(script);


twitter_autoload.readable.js
/*! Twitter autoload source | (c) 2016 William Wendell */

(function(){
	'use strict';
	var Autoload = function(){
		
		this.uq=false; // update queued
		this.v=!document.hidden; // page visible
		this.exp = -1; // delay exponent
		this.to = false; // timeout reference

		// new tweets bar observer
		this.o = new MutationObserver((function(m){  
			this.undelay('new tweets bar!');
		}).bind(this));
		
		// tweets observer
		this.o2 = new MutationObserver((function(m){ 
			this.undelay('feed changed'); 
		}).bind(this));
		
		// observer config
		this.oc = { attributes: false, childList: true, characterData: false };
		this.tl = document.getElementById('stream-items-id');
		this.t0 = 2500; // initial polling delay

		document.addEventListener('visibilitychange', 
			(function(ev){ 
				if (this.v=!document.hidden && this.uq) { 
					console.log('we´re visible again');
					this.tk(); 
				} 
			}).bind(this), false
		); 
		this.o.observe(document.querySelector('.js-new-items-bar-container'), this.oc);
		this.o2.observe(this.tl, this.oc);
		
		this.keepup();
	};

	Object.assign(Autoload.prototype, {
		su: function(px) {
			function getOffsetSum(e) {
				var top=0, left=0;
				while(e) {
					top = top + parseInt(e.offsetTop);
					left = left + parseInt(e.offsetLeft);
					e = e.offsetParent;
				}
				return {top: top, left: left};
			}

			return !document.querySelectorAll('input:focus,textarea:focus,[contenteditable=\"true\"]:focus').length
				&& window.pageYOffset - document.documentElement.clientTop - getOffsetSum(this.tl).top < px
				&& this.v;
		},
		tk: function() {
			if (this.su(15)) { 
				document.dispatchEvent(new CustomEvent('uiShortcutGotoTopOfScreen'));
				this.uq = false;
			} else {
				if (this.v) { 
					console.log('waiting...');
					window.setTimeout( this.tk.bind(this), 100); 
				} else {
					console.log('exit tk loop, page hidden');
				}
			}
		},
		up: function() {
			if (!this.uq) {
				this.uq = true;
				this.tk();
				return true;
			} else {
				return false;
			}
		},
		undelay: function(f) {
			console.log('undelayed: '+f)
			this.exp = -1;
			window.clearTimeout(this.to);
			this.to = window.setTimeout(this.keepup.bind(this), 100 ); 
		},
		delay: function() {
			if (this.exp < 10) { ++this.exp; }
		},
		keepup: function() {
			if (this.up()) {
				this.delay();
			}
			console.log('timeout: '+parseInt(this.t0 * Math.pow(1.3, this.exp)));
			this.to = window.setTimeout(this.keepup.bind(this), parseInt(this.t0 * Math.pow(1.3, this.exp)) ); 
		}
	});
	
	
	window.onload = window.setTimeout(new Autoload(), 5000);
})();
